timemachine/cpp/src/barostat.cu

```diff
@@ -0,0 +1,295 @@
+
+    gpuErrchk(cudaMalloc(&d_u_before_, N_*sizeof(*d_u_before_)));
+    gpuErrchk(cudaMalloc(&d_u_after_, N_*sizeof(*d_u_after_)));
+
+    gpuErrchk(cudaMalloc(&d_centroids_, num_mols*3*sizeof(*d_centroids_)));
+    gpuErrchk(cudaMalloc(&d_mol_idxs_, N_*sizeof(*d_mol_idxs_)));
+    gpuErrchk(cudaMalloc(&d_mol_offsets_, (num_mols+1)*sizeof(*d_mol_offsets_)));
+
+    int offset = 0;
+    int mol_offsets[num_mols+1];
+    int mol_idxs[N_];
+    for (int i = 0; i < num_mols; i++) {
+        // Sort the array in ascending order
+        mol_offsets[i] = offset;
+        int num_atoms = group_idxs[i].size();
+        for (int j = 0; j < num_atoms; j++) {
+            mol_idxs[offset+j] = i;
+        }
+        offset += group_idxs[i].size();
+    }
+    mol_offsets[num_mols] = offset;
+    gpuErrchk(cudaMemcpy(d_mol_idxs_, mol_idxs, N_*sizeof(*d_mol_idxs_), cudaMemcpyHostToDevice));
+    gpuErrchk(cudaMemcpy(d_mol_offsets_, mol_offsets, (num_mols+1)*sizeof(*d_mol_offsets_), cudaMemcpyHostToDevice));
+
+};
+
+MonteCarloBarostat::~MonteCarloBarostat() {
+    gpuErrchk(cudaFree(d_x_after_));
+    gpuErrchk(cudaFree(d_centroids_));
+    gpuErrchk(cudaFree(d_mol_idxs_));
+    gpuErrchk(cudaFree(d_box_after_));
+    gpuErrchk(cudaFree(d_u_before_));
+    gpuErrchk(cudaFree(d_u_after_));
+};
+
+
+void __global__ rescale_positions(
+    const int N, // Number of coords
+    double * __restrict__ coords, // Cordinates [N*3]
+    const double length_scale,
+    const double * __restrict__ box, // [9]
+    double * __restrict__ scaled_box, // [9]
+    const int * __restrict__ mol_idxs, // [N]
+    const int * __restrict__ mol_offsets,
+    const double * __restrict__ centroids // [N*3]
+) {
+    const int atom_idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (atom_idx >= N) {
+        return;
+    }
+    const int mol_idx = mol_idxs[atom_idx];
+    double center_x = box[0*3+0] * 0.5;
+    double center_y = box[1*3+1] * 0.5;
+    double center_z = box[2*3+2] * 0.5;
+
+    const double num_atoms = static_cast<double>(mol_offsets[mol_idx+1] - mol_offsets[mol_idx]);
+
+    const double centroid_x = centroids[mol_idx*3+0] / num_atoms;
+    const double centroid_y = centroids[mol_idx*3+1] / num_atoms;
+    const double centroid_z = centroids[mol_idx*3+2] / num_atoms;
+
+    const double displacement_x = ((centroid_x - center_x) * length_scale) + center_x - centroid_x;
+    const double displacement_y = ((centroid_y - center_y) * length_scale) + center_y - centroid_y;
+    const double displacement_z = ((centroid_z - center_z) * length_scale) + center_z - centroid_z;
+
+    coords[atom_idx*3+0] += displacement_x;
+    coords[atom_idx*3+1] += displacement_y;
+    coords[atom_idx*3+2] += displacement_z;
+    if (atom_idx == 0) {
+        scaled_box[0*3+0] *= length_scale;
+        scaled_box[1*3+1] *= length_scale;
+        scaled_box[2*3+2] *= length_scale;
+    }
+}
+
+void __global__ find_group_centroids(
+    const int N, // Number of coords
```

Comment:
Are you sure this should be number of coords as opposed to sum of group_idxs size?

Reply:
With latest changes it is still N, but for mol idxs that are invalid the value is set to -1